File: tests/Tests.FSharp/Formal/Tlc.Attempts.fs
module Zeta.Tests.Formal.TlcAttempts

open System
open System.Diagnostics
open System.IO
open System.Security.Cryptography
open System.Text.Json
open System.Text.RegularExpressions
open System.Threading
open System.Threading.Tasks

type FileIdentity = { File: string; Bytes: int64; Sha256: string }
type InputIdentity = { Source: FileIdentity; Copied: FileIdentity }
type Attempt =
    { Directory: string; Workspace: string; Metadir: string
      Stdout: string; Stderr: string; ErrorFile: string; Inputs: InputIdentity[] }
type PreparationFailure = { Directory: string; Error: string }
type ProcessCapture = { ExitCode: int; TimedOut: bool }

/// Start and drain one real process under one caller-owned cancellation source.
/// The caller owns the process, streams and token until this operation completes.
let captureProcessWithDeadline (proc: Process) (stdoutFile: FileStream) (stderrFile: FileStream) (deadline: CancellationTokenSource) =
    if not (proc.Start()) then invalidOp "process did not start"
    let copyOut = proc.StandardOutput.BaseStream.CopyToAsync(stdoutFile, deadline.Token)
    let copyErr = proc.StandardError.BaseStream.CopyToAsync(stderrFile, deadline.Token)
    let exit = proc.WaitForExitAsync deadline.Token
    let mutable timedOut = false
    try Task.WhenAll([|exit; copyOut; copyErr|]).GetAwaiter().GetResult()
    with :? OperationCanceledException when deadline.IsCancellationRequested ->
        // One deadline covers both the direct child and inherited-pipe EOF.
        // Awaiting WhenAll also observes both cancelled copy tasks before disposing files.
        timedOut <- true
        if not proc.HasExited then
            try proc.Kill true
            with :? InvalidOperationException when proc.HasExited -> ()
        proc.WaitForExit()
    stdoutFile.Flush true
    stderrFile.Flush true
    { ExitCode = proc.ExitCode; TimedOut = timedOut }

/// File ownership survives failed launch or second-stream opening. The optional
/// deadline is for metadata probes; the F# model checker retains its existing no-timeout policy.
let captureProcess executable (argv: string seq) cwd stdout stderr (timeout: int option) =
    let info = ProcessStartInfo(executable)
    info.WorkingDirectory <- cwd
    info.UseShellExecute <- false
    info.RedirectStandardOutput <- true
    info.RedirectStandardError <- true
    for argument in argv do info.ArgumentList.Add argument
    use stdoutFile = new FileStream(stdout, FileMode.CreateNew, FileAccess.Write, FileShare.Read)
    use stderrFile = new FileStream(stderr, FileMode.CreateNew, FileAccess.Write, FileShare.Read)
    use deadline = new CancellationTokenSource()
    match timeout with
    | Some milliseconds -> deadline.CancelAfter milliseconds
    | None -> ()
    use proc = new Process(StartInfo = info)
    captureProcessWithDeadline proc stdoutFile stderrFile deadline

let identifyFile name path =
    use stream = File.OpenRead path
    { File = name; Bytes = stream.Length; Sha256 = Convert.ToHexString(SHA256.HashData stream) }

let private generatedInput (name: string) =
    name.Contains("_TTrace_", StringComparison.Ordinal)
    || (name.StartsWith("MC", StringComparison.Ordinal) && name.EndsWith(".tla", StringComparison.Ordinal))

/// Include unstaged/new sources; refuse ignored local helpers rather than omit them.
let sourceInputs root specsPath (required: string[]) =
    let info = ProcessStartInfo("git")
    info.WorkingDirectory <- root
    info.UseShellExecute <- false
    info.RedirectStandardOutput <- true
    info.RedirectStandardError <- true
    for argument in ["ls-files"; "--cached"; "--others"; "--exclude-standard"; "--"; "src/Core.TLA/specs/*.tla"; "src/Core.TLA/specs/*.cfg"] do
        info.ArgumentList.Add argument
    use proc = Process.Start info
    let stdout = proc.StandardOutput.ReadToEnd()
    let stderr = proc.StandardError.ReadToEnd()
    proc.WaitForExit()
    if proc.ExitCode <> 0 then invalidOp ("git source inventory refused: " + stderr)
    let paths = stdout.Split([|'\n'; '\r'|], StringSplitOptions.RemoveEmptyEntries)
    if paths |> Array.exists (fun path -> path <> "src/Core.TLA/specs/" + Path.GetFileName path) then
        invalidOp "nested or quoted source path is unsupported"
    let names = paths |> Array.map Path.GetFileName |> Array.filter (generatedInput >> not)
    for name in required do
        if not (Array.contains name names) then invalidOp ("selected input absent from git-visible source closure: " + name)
    for file in Directory.EnumerateFiles specsPath do
        let name = Path.GetFileName file
        if (name.EndsWith(".tla", StringComparison.Ordinal) || name.EndsWith(".cfg", StringComparison.Ordinal))
           && not (generatedInput name) && not (Array.contains name names) then
            invalidOp ("ignored local source input would be omitted: " + name)
    names

/// Startup evidence is vetoed by any checker execution or fatal/in-run JVM marker.
let jvmNeverStarted (output: string) =
    let startedOrFatal = @"TLC2|TLC Version|Starting\.\.\.|Computing initial|Finished computing|Progress\(|distinct states|Model checking|Invariant .*violated|TLC bug|SIG[A-Z]+|fatal error|hs_err|OutOfMemoryError"
    let startup = @"Error occurred during initialization of VM|Could not reserve enough space for (?:\S+ )?object heap|Unable to access jarfile|Could not create the Java Virtual Machine"
    not (Regex.IsMatch(output, startedOrFatal, RegexOptions.IgnoreCase ||| RegexOptions.CultureInvariant))
    && Regex.IsMatch(output, startup, RegexOptions.IgnoreCase ||| RegexOptions.CultureInvariant)

let private writeNew path value =
    use stream = new FileStream(path, FileMode.CreateNew, FileAccess.Write, FileShare.Read)
    JsonSerializer.Serialize(stream, value, JsonSerializerOptions(WriteIndented = true))
    stream.Flush true

/// Copy the whole admitted local .tla/.cfg closure into an attempt-owned workspace.
/// A copy refusal leaves its directory and explicit preparation failure intact.
let prepare diagnosticsRoot model attempt specsPath (sourceInventory: unit -> string[]) =
    let mutable directory = ""
    let copiedInputs = ResizeArray<InputIdentity>()
    try
        Directory.CreateDirectory diagnosticsRoot |> ignore
        let safeModel = Regex.Replace(model, "[^a-zA-Z0-9_-]", "_")
        directory <- Path.Combine(diagnosticsRoot, safeModel + "-" + Guid.NewGuid().ToString("N"))
        Directory.CreateDirectory directory |> ignore
        writeNew (Path.Combine(directory, "attempt.json")) {| Stage = "preparation"; Model = model; Attempt = attempt |}
        let workspace = Path.Combine(directory, "workspace")
        let metadir = Path.Combine(directory, "states")
        Directory.CreateDirectory workspace |> ignore
        Directory.CreateDirectory metadir |> ignore
        let sources = sourceInventory()
        if sources.Length = 0 || (Array.distinct sources |> Array.length) <> sources.Length then
            invalidArg (nameof sources) "source inventory must be nonempty and unique"
        let inputs =
            sources |> Array.sort |> Array.map (fun name ->
                if Path.GetFileName name <> name
                   || not (name.EndsWith(".tla", StringComparison.Ordinal) || name.EndsWith(".cfg", StringComparison.Ordinal))
                   || name.Contains("_TTrace_", StringComparison.Ordinal)
                   || (name.StartsWith("MC", StringComparison.Ordinal) && name.EndsWith(".tla", StringComparison.Ordinal)) then
                    invalidArg (nameof sources) ("refusing non-source input: " + name)
                let original = Path.Combine(specsPath, name)
                if File.GetAttributes original &&& FileAttributes.ReparsePoint <> enum 0 then
                    invalidArg (nameof sources) ("refusing linked source input: " + name)
                let copied = Path.Combine(workspace, name)
                File.Copy(original, copied, false)
                let sourceIdentity = identifyFile name original
                let copiedIdentity = identifyFile name copied
                if sourceIdentity <> copiedIdentity then invalidOp ("source changed while copied: " + name)
                let input = { Source = sourceIdentity; Copied = copiedIdentity }
                copiedInputs.Add input
                input)
        Ok { Directory = directory; Workspace = workspace; Metadir = metadir; Inputs = inputs
             Stdout = Path.Combine(directory, "stdout.log"); Stderr = Path.Combine(directory, "stderr.log")
             ErrorFile = Path.Combine(directory, "hs_err_pid%p.log") }
    with error ->
        if directory <> "" then
            try writeNew (Path.Combine(directory, "preparation-failure.json")) {| Stage = "preparation"; Error = error.Message; CopiedInputs = copiedInputs.ToArray() |}
            with _ -> () // Report the original path even when the failure receipt cannot be written.
        Error { Directory = directory; Error = error.Message }

/// Unexpected state directories stay intact. This counts bytes, not a retention quota.
let inventory directory =
    Directory.EnumerateFiles(directory, "*", SearchOption.AllDirectories)
    |> Seq.fold (fun (files, bytes) path -> files + 1, bytes + FileInfo(path).Length) (0, 0L)

let writeDiagnostic (attempt: Attempt) (name: string) value =
    if not (Regex.IsMatch(name, "^[a-z-]+\\.json$", RegexOptions.CultureInvariant)) then
        invalidArg (nameof name) "diagnostic filename must be an owned JSON basename"
    writeNew (Path.Combine(attempt.Directory, name)) value

/// Invoke only after the entire semantic judgement, including expected violations.
let finish (attempt: Attempt) expected =
    if expected then Directory.Delete(attempt.Directory, true)

File: src/Core.TypeScript/formal-verification/tlc-attempts.ts
// TLC diagnostic ownership and retry policy. No JVM/model policy lives here.
import {
  closeSync, constants, fstatSync, lstatSync, mkdirSync, mkdtempSync, openSync, readFileSync, readdirSync,
  rmSync, statSync, writeFileSync,
} from "node:fs";
import { createHash } from "node:crypto";
import { basename, join } from "node:path";
import { spawnSync, type SpawnSyncReturns } from "node:child_process";

export interface AttemptDirectory {
  readonly directory: string;
  readonly workspace: string;
  readonly metadir: string;
  readonly stdout: string;
  readonly stderr: string;
  readonly errorFile: string;
  readonly inputs: readonly SourceInputIdentity[];
}

export interface FileIdentity {
  readonly File: string;
  readonly Bytes: number;
  readonly Sha256: string;
}

export interface SourceOpenPolicy {
  readonly Flags: number;
  readonly NoFollowAvailable: boolean;
  readonly NonBlockingAvailable: boolean;
}

export interface SourceInputIdentity extends FileIdentity {
  readonly CopiedSha256: string;
  readonly OpenPolicy: SourceOpenPolicy;
}

/** Optional POSIX flags strengthen final-component admission. Their absence
 * keeps the stable writer-tree contract; it does not silently remove Windows. */
export function sourceOpenPolicy(available: { O_NOFOLLOW?: number; O_NONBLOCK?: number } = constants): SourceOpenPolicy {
  const noFollow = available.O_NOFOLLOW ?? 0;
  const nonBlocking = available.O_NONBLOCK ?? 0;
  return { Flags: constants.O_RDONLY | noFollow | nonBlocking,
    NoFollowAvailable: noFollow !== 0, NonBlockingAvailable: nonBlocking !== 0 };
}

/** Borrow one opened descriptor: type admission and bytes refer to that same
 * file even if its pathname is replaced. The caller owns descriptor closure. */
export function copyRegularSourceDescriptor(descriptor: number, destination: string, name: string): FileIdentity & { CopiedSha256: string } {
  if (!fstatSync(descriptor).isFile()) throw new Error("source input is not a regular file: " + name);
  const bytes = readFileSync(descriptor);
  const original = { File: name, Bytes: bytes.length, Sha256: createHash("sha256").update(bytes).digest("hex").toUpperCase() };
  writeFileSync(destination, bytes, { flag: "wx" });
  const copied = identifyFile(destination, name);
  if (original.Sha256 !== copied.Sha256) throw new Error("captured source/copy bytes differ: " + name);
  return { ...original, CopiedSha256: copied.Sha256 };
}

export type DiagnosticResult<T> =
  | { readonly ok: true; readonly value: T }
  | { readonly ok: false; readonly error: string; readonly directory: string };

export function identifyFile(path: string, name: string = path): FileIdentity {
  const bytes = readFileSync(path);
  return { File: name, Bytes: bytes.length, Sha256: createHash("sha256").update(bytes).digest("hex").toUpperCase() };
}

/** Both descriptors are owned even if opening the second one fails. Raw streams
 * bypass spawn's memory-buffer ceiling. Timeout applies only to this subprocess. */
export function captureProcess(
  executable: string, argv: readonly string[], cwd: string,
  stdout: string, stderr: string, timeout: number, killSignal?: NodeJS.Signals,
): SpawnSyncReturns<Buffer> {
  let output = -1;
  let errorOutput = -1;
  try {
    output = openSync(stdout, "wx");
    errorOutput = openSync(stderr, "wx");
    const result = spawnSync(executable, [...argv], {
      cwd, stdio: ["ignore", output, errorOutput], timeout,
      ...(killSignal === undefined ? {} : { killSignal }),
    });
    // Bun may omit status/signal on launch failure; receipts use explicit nulls.
    return { ...result, status: result.status ?? null, signal: result.signal ?? null };
  } finally {
    if (output >= 0) closeSync(output);
    if (errorOutput >= 0) closeSync(errorOutput);
  }
}

const generatedInput = (name: string): boolean => /_TTrace_|^MC.*\.tla$/.test(name);

/** Admit new unstaged sources too. Ignored non-generated inputs are an explicit
 * refusal, since silently omitting a local helper changes module resolution. */
export function sourceInputs(root: string, specsPath: string, required: readonly string[]): readonly string[] {
  const found = spawnSync("git", ["ls-files", "--cached", "--others", "--exclude-standard", "--",
    "src/Core.TLA/specs/*.tla", "src/Core.TLA/specs/*.cfg"], { cwd: root, encoding: "utf8" });
  if (found.status !== 0) throw new Error("git source inventory refused: " + String(found.error ?? found.stderr));
  const paths = found.stdout.split("\n").filter((path) => path !== "");
  if (paths.some((path) => path !== "src/Core.TLA/specs/" + basename(path))) throw new Error("nested or quoted source path is unsupported");
  const names = paths.map((path) => basename(path)).filter((name) => !generatedInput(name));
  for (const name of required) if (!names.includes(name)) throw new Error("selected input absent from git-visible source closure: " + name);
  for (const name of readdirSync(specsPath)) {
    if (/\.(tla|cfg)$/.test(name) && !generatedInput(name) && !names.includes(name)) {
      throw new Error("ignored local source input would be omitted: " + name);
    }
  }
  return names;
}

/** Conservative startup evidence. A checker banner, progress or fatal signal
 * takes precedence even if the same output also quotes a startup marker. */
export function jvmNeverStarted(stdout: string, stderr: string = ""): boolean {
  const output = stdout + "\n" + stderr;
  const startedOrFatal = /TLC2|TLC Version|Starting\.\.\.|Computing initial|Finished computing|Progress\(|distinct states|Model checking|Invariant .*violated|TLC bug|SIG[A-Z]+|fatal error|hs_err|OutOfMemoryError/i;
  const startup = /Error occurred during initialization of VM|Could not reserve enough space for (?:\S+ )?object heap|Unable to access jarfile|Could not create the Java Virtual Machine/i;
  return !startedOrFatal.test(output) && startup.test(output);
}

export interface AttemptOutcome {
  readonly ok: boolean;
  readonly exitCode: number;
  readonly signal: string | null;
  readonly processError: boolean;
  readonly stdout: string;
  readonly stderr: string;
}

/** Retain each result, including failures preceding a startup recovery. */
export function runWithStartupRetry<T extends AttemptOutcome>(
  run: (attempt: number) => T,
  settle: (attempt: number) => void,
): readonly T[] {
  const results: T[] = [];
  for (let attempt = 1; attempt <= 3; attempt++) {
    const result = run(attempt);
    results.push(result);
    if (result.ok || attempt === 3 || result.exitCode !== 1 || result.signal !== null || result.processError
      || !jvmNeverStarted(result.stdout, result.stderr)) break;
    settle(attempt);
  }
  return results;
}

/** Only admitted source basenames are copied. Symlink observation assumes a
 * stable writer-owned tree on platforms without no-follow. This is not hostile
 * ancestor isolation or an atomic snapshot of concurrently modified files. */
export function prepareAttempt(
  diagnosticsRoot: string,
  model: string,
  attempt: number,
  specsPath: string,
  sourceInventory: readonly string[] | (() => readonly string[]),
): DiagnosticResult<AttemptDirectory> {
  let directory = "";
  const inputs: SourceInputIdentity[] = [];
  try {
    mkdirSync(diagnosticsRoot, { recursive: true });
    directory = mkdtempSync(join(diagnosticsRoot, model.replace(/[^a-zA-Z0-9_-]/g, "_") + "-"));
    writeFileSync(join(directory, "attempt.json"), JSON.stringify({ Stage: "preparation", Model: model, Attempt: attempt, SourceOpenPolicy: sourceOpenPolicy() }) + "\n", { flag: "wx" });
    const workspace = join(directory, "workspace");
    const metadir = join(directory, "states");
    mkdirSync(workspace);
    mkdirSync(metadir);
    const sources = typeof sourceInventory === "function" ? sourceInventory() : sourceInventory;
    if (sources.length === 0 || new Set(sources).size !== sources.length) throw new Error("source inventory must be nonempty and unique");
    for (const name of [...sources].sort()) {
      if (basename(name) !== name || !/\.(tla|cfg)$/.test(name) || /_TTrace_|^MC.*\.tla$/.test(name)) {
        throw new Error("refusing non-source input: " + name);
      }
      const path = join(specsPath, name);
      const policy = sourceOpenPolicy();
      const descriptor = openSync(path, policy.Flags);
      try {
        // Static symlink observation also covers platforms without O_NOFOLLOW.
        // It never authorizes a second source-path open: consumption uses the fd.
        if (!lstatSync(path).isFile()) throw new Error("source path is not an observed regular file: " + name);
        inputs.push({ ...copyRegularSourceDescriptor(descriptor, join(workspace, name), name), OpenPolicy: policy });
      } finally { closeSync(descriptor); }
    }
    return { ok: true, value: {
      directory, workspace, metadir, inputs,
      stdout: join(directory, "stdout.log"), stderr: join(directory, "stderr.log"),
      errorFile: join(directory, "hs_err_pid%p.log"),
    } };
  } catch (error) {
    const detail = String(error);
    if (directory !== "") {
      try { writeFileSync(join(directory, "preparation-failure.json"), JSON.stringify({ Stage: "preparation", Error: detail, CopiedInputs: inputs }) + "\n", { flag: "wx" }); }
      catch { /* The caller still reports the original path and write failure. */ }
    }
    return { ok: false, error: detail, directory };
  }
}

/** Inventory ownership, not a disk quota: unexpected state bytes are never
 * silently capped or purged. Local failed directories may be large. */
export function inventory(directory: string): { Files: number; Bytes: number } {
  let files = 0;
  let bytes = 0;
  for (const entry of readdirSync(directory, { withFileTypes: true })) {
    const path = join(directory, entry.name);
    if (entry.isDirectory()) {
      const child = inventory(path);
      files += child.Files;
      bytes += child.Bytes;
    } else if (entry.isFile()) {
      files++;
      bytes += statSync(path).size;
    }
  }
  return { Files: files, Bytes: bytes };
}

export function writeDiagnostic(attempt: AttemptDirectory, name: string, value: unknown): void {
  if (!/^[a-z-]+\.json$/.test(name)) throw new Error("diagnostic filename is not an owned JSON basename");
  writeFileSync(join(attempt.directory, name), JSON.stringify(value, null, 2) + "\n", { flag: "wx" });
}

/** Called only after all semantic verdict checks, including expected violations. */
export function finishAttempt(attempt: AttemptDirectory, expected: boolean): void {
  if (expected) rmSync(attempt.directory, { recursive: true });
}
